lib/http_aws_sigv4.c

```diff
@@ -66,6 +68,159 @@ static void sha256_to_hex(char *dst, unsigned char *sha, size_t dst_l)
   }
 }
 
+
+
+static CURLcode make_headers(struct Curl_easy *data,
+                             const char *hostname,
+                             const char *timestamp,
+                             const char *provider1_low,
+                             const char *provider1_mid,
+                             char **canonical_headers,
+                             char **signed_headers)
+{
+  char *date_hdr_key =  curl_maprintf("X-%s-Date", provider1_mid);
+  char *date_full_hdr = curl_maprintf("x-%s-date:%s",
+                                      provider1_low, timestamp);
+#ifdef DEBUGBUILD
+  char *force_host = getenv("CURL_FORCEHOST");
+#endif
+  struct curl_slist *head = NULL;
+  struct curl_slist *tmp_head = NULL;
+  int ret = CURLE_OUT_OF_MEMORY;
+  struct curl_slist *l;
+  int again = 1;
+
+  if(!date_hdr_key || !date_full_hdr)
+    goto fail;
+
+  if(Curl_checkheaders(data, "Host")) {
+    head = NULL;
+  }
+#ifdef DEBUGBUILD
+  else if(force_host) {
+    head = curl_slist_append(NULL, "host:127.0.0.1");
+    if(!head)
+      goto fail;
+  }
+#endif
+  else {
+    char *full_host = data->state.aptr.host ? data->state.aptr.host :
+      curl_maprintf("host:%s", hostname);
+
+    if(!full_host)
+      goto fail;
+    head = curl_slist_append(NULL, full_host);
+    if(full_host != data->state.aptr.host)
+      free(full_host);
+    if(!head)
+      goto fail;
+  }
+
+
+  if(!date_hdr_key || !date_full_hdr)
+    goto fail;
+
+  for(l = data->set.headers; l; l = l->next) {
+    tmp_head = curl_slist_append(head, l->data);
+    if(!tmp_head)
+      goto fail;
+    head = tmp_head;
+  }
+
+  /* remove whitespace, and lowercase each headers */
+  for(l = head; l; l = l->next) {
+    char *tmp = l->data, *tmp2;
+
+    Curl_strntolower(l->data, l->data, strlen(l->data));
+    tmp = strchr(tmp, ':');
+    if(!tmp)
+      continue;
+    ++tmp;
+    tmp2 = tmp;
+    while(*tmp2 == ' ' || *tmp2 == '\t') {
+      ++tmp2;
+    }
+    if(tmp != tmp2)
+      memmove(tmp, tmp2, strlen(tmp2) + 1);
+
+    while(*tmp) {
+      tmp2 = tmp;
+      while(*tmp2 == ' ' || *tmp2 == '\t') {
+        ++tmp2;
+      }
+
+      if(tmp != tmp2) {
+        if(*tmp2 == 0)
+          /* remove spaces at the end */
+          *tmp = 0;
+        else {
+          *tmp++ = ' '; /* if tab convert to space */
+          if (*tmp == ' ' || *tmp == '\t')
+            memmove(tmp, tmp2, strlen(tmp2));
+        }
```

Comment:
I may be wrong here, but the spec says:
```
Trim() Remove any leading or trailing white-space. 
```
and this code seems to do extra work on intermediate spaces. Do you know any valid examples with space-containing header values?

Please note, that trailing `\r\n` (from `data->state.aptr.host`) also could be trimmed here.

Reply:
yes, your right, I'm presently working on fixing this, thanks for the feedback.

Reply:
The `Host:` bug seems to be fixed now.